zipline/gens/tradesimulation.py

```diff
@@ -186,51 +190,135 @@ def once_a_day(midnight_dt):
-
-            # if there was a daily perf message, collect and yield it
-            if daily_message:
-                daily_message['daily_perf']['recorded_vars'] = rvars
-                yield daily_message
+        if sid is None:
+            # get benchmark info from trading environment
+            benchmark_series = \
+                env.benchmark_returns[trading_days[0]:trading_days[-1]]
+        else:
+            # check if this security has a stock dividend.  if so, raise an
+            # error suggesting that the user pick a different asset to use
+            # as benchmark.
+            stock_dividends = \
+                data_portal.get_stock_dividends(sid, trading_days)
+
+            if len(stock_dividends) > 0:
+                raise InvalidBenchmarkAsset(
+                    sid=str(sid),
+                    dt=stock_dividends[0]["ex_date"]
+                )
+
+            benchmark_asset = env.asset_finder.retrieve_asset(sid)
+            if benchmark_asset.start_date > trading_days[0]:
+                # the asset started trading after the first simulation day
+                raise BenchmarkAssetNotAvailableTooEarly(
+                    sid=str(sid),
+                    dt=trading_days[0],
+                    start_dt=benchmark_asset.start_date
+                )
+
+            if benchmark_asset.end_date < trading_days[-1]:
+                # the asset stopped trading before the last simulation day
+                raise BenchmarkAssetNotAvailableTooLate(
+                    sid=str(sid),
+                    dt=trading_days[0],
+                    end_dt=benchmark_asset.end_date
+                )
+
+            # get the window of close prices for benchmark_sid from the last
+            # trading day of the simulation, going up to one day before the
+            # simulation start day.
+            benchmark_series = data_portal.get_history_window(
+                [sid],
+                trading_days[-1],
+                bar_count=len(trading_days) + 1,
+                frequency="1d",
+                field="close"
+            )[sid]
+
+            # now, we need to check if we can safely go use the
+            # one-day-before-sim-start value, by seeing if the asset was
+            # trading that day.
+            trading_day_before_sim_start = \
+                env.previous_trading_day(trading_days[0])
+
+            if benchmark_asset.start_date > trading_day_before_sim_start:
+                # we can't go back one day before sim start, because the asset
+                # didn't start trading until the same day as the sim start.
+                # instead, we'll use the first available minute value of the
+                # first sim day.
+                minutes_in_first_day = \
+                    env.market_minutes_for_day(trading_days[0])
+
+                # get a minute history window of the first day
+                minute_window = data_portal.get_history_window(
```

Comment:
This works, I can't recall if we needed a data_portal "get spot price with ffill" elsewhere. We could use that type of method here instead.